### PR TITLE
Add single quote handling for unescaped strings

### PR DESCRIPTION
**Pull request by ansiwen**
Parse_argv handles `\` and `"` as special characters in any context, which makes the usage of boot parameters containing these characters &ndash; like JSON strings &ndash; very tedious.  This change introduces the handling of single quoted text ('foobar'), that - like in typical shell command line parsing - is interpreted as verbatim text without any special escape handling.  The only character that has special handling is the single quote `'` itself, which ends the single quoted text and therefore can't appear within single quoted text. (Also `\'` is not supported.) Now for example this command:
```
$ ./solo5-hvt --net=tap100 ./app.hvt --json '{\"foo\":\ \"bar\"}'
```
can instead be written as
```
$ ./solo5-hvt --net=tap100 ./app.hvt --json \''{"foo": "bar"}'\'
```

Related: Solo5/solo5#281
Related: Solo5/solo5#283
Related: mirage/mirage-solo5#33

### CODE REVIEW

**Comment by samoht** [comment]
Will `./solo5-hvt --net=tap100 ./app.hvt --json '{"foo": "bar"}'` be parsed correctly?

**Comment by ansiwen** [comment]
> Will `./solo5-hvt --net=tap100 ./app.hvt --json '{"foo": "bar"}'` be parsed correctly?

@samoht No, neither before nor after this PR it would be parsed correctly (depending on what you regard as "correctly"). The `'` are consumed by the shell already, and the `"` then would be consumed by Parse_argv, resulting in an option value of `{foo: bar}` which is not a valid json anymore. The only way to make this work would be to add escaping within the solo5 tender, but @hannes and @mato were not in favor of that idea, see discussion here: mirage/mirage-solo5#33

**Comment by mato** [comment]
@ansiwen Stepping back a bit here, passing _JSON_ on a "command line" (be that shell, QEMU, xl.cfg or whatever else) is generally a bad idea. BTDT, lots of pain involved.

MirageOS already provides configuration keys/values via Functoria. Is there any particular reason the information you want to pass to your unikernel cannot be represented by key/values as provided today?

**Comment by ansiwen** [comment]
> @ansiwen Stepping back a bit here, passing _JSON_ on a "command line" (be that shell, QEMU, xl.cfg or whatever else) is generally a bad idea. BTDT, lots of pain involved.

@mato I don't think this is about JSON or whatever data format, but about the (non-existing) possibility to forward verbatim strings to the unikernel as parameter. So, which characters in boot parameters do you consider safe, which not?

> MirageOS already provides configuration keys/values via Functoria. Is there any particular reason the information you want to pass to your unikernel cannot be represented by key/values as provided today?

I don't understand this. I _am_ using Functoria. What I was referring to as "option value" in my previous comment is what you get from Functoria when I would get the option value via `Key_gen.json ()`. But maybe you are referring to something else in Functoria, that I haven't discovered yet?

**Edit:** Ok, now I think I understand, you mean directly using `--foo bar` as boot parameter? Well, in this case it is an arbitrary text file, that the webserver has to serve, and happens to be in json. I don't have control over the structure of the content, it's part of a protocol. So, even _if_ I knew the inner structure of the JSON and would map them to individual boot parameters (but then, what would I do with nested maps and arrays?), I would have to reconstruct the JSON, because that's what the service has to deliver. Just forget about JSON, see it as an arbitrary text, that is just passed through from the command line through the webserver to the client of the service, which is the only party that knows what to do with this text.

**Comment by samoht** [comment]
I am wondering if it would help if we document how to escape easily a string in the shell; for instance: 

```
./solo5-hvt --net=tap100 ./app.hvt --json $(printf "%q" '{"foo": "bar"}')
```

and/or we could ship a simple binary which does the right escaping; Note that I still don't have yet a global understanding of the issue here, so please bear with me if I am causing too much noise :p

**Comment by hannesm** [comment]
looking again at this issue, I don't understand the modification, as in what is supported and what not? I can see that some test cases are added, but what happens in nested quote modes (i.e. `"'foo'"` and `'"foo"'`)?

I'm not an expert in command-line arguments, for me that currently supported by MirageOS/parse-argv are sufficient, but I understand that escaping using backslashes is not good for all the use cases. maybe someone from the @mirage/core team interested in command-line parsing can take a deeper look? from what I understand, I now additionally need to escape my `'` in command-line arguments, but as an exchange I get `'` handling, where `"` and whitespaces may appear - is this correct?

if others are in favour of this PR, and before merging, I'd suggest:
- regarding the test cases, they are pretty hard to read due to OCaml's string escaping syntax, I'd suggest to revise them using `{|foo|}`, since there's no need to escape `"` inside of that construct.
- given that this PR changes the semantics of parse-argv drastically, I'd suggest to also document the (new) semantics in the `mli` file, and summarize it in README.md.

**Comment by mato** [comment]
@ansiwen:

> @mato I don't think this is about JSON or whatever data format, but about the (non-existing) possibility to forward verbatim strings to the unikernel as parameter. So, which characters in boot parameters do you consider safe, which not?

As a conservative guess, over all the backends? `[A-Za-z0-9_]`, which is probably not what you wanted to hear.

> Ok, now I think I understand, you mean directly using --foo bar as boot parameter? Well, in this case it is an arbitrary text file, that the webserver has to serve, and happens to be in json.

(sorry, did not see this, it got added as an edit and I don't get notifications for that).

Yes, this is what I meant. Ignoring the JSON question, what I think what you're asking for is:

* Is there a structured way to pass arbitrary items of _data_ to a unikernel at boot time?

Where _data_ is something less restrictive than "command-line safe characters", i.e. in the general case could just be "binary blob", or at least "text/plain".

My argument is that the "command line" is the wrong abstraction for this. I agree that we don't have a good alternative, however, rather than trying to shoe-horn this into the "command line", perhaps we should instead consider the above as a requirement for an enhancement and design a new abstraction, taking into account the various backends and how they are booted?

**Comment by mato** [comment]
> should instead consider the above as a requirement for an enhancement 

Oh, and thinking about it, there is a simple if not super efficient solution to your use case: Just pass the data in as Base64, as defined in [RFC 4648](https://tools.ietf.org/html/rfc4648). I.e. `--json=<Base64>` and then decode it on the unikernel side. Problem solved. Note that there are some (undocumented) limits for command line length in Solo5, but I presume you're not trying to pass something larger than a couple of kB.

**Comment by samoht** [comment]
We could also add a `file` datatype where you want to pass the contents of a file to the unikernel (e.g. `--foo=file.json`). I think this could be a nice addition (and the contents of the file could be automatically base-64 encoded by the mirage tool to avoid annoying escaping).

**Comment by ansiwen** [comment]
Sorry as well for my long silence.

@samoht:
> I am wondering if it would help if we document how to escape easily a string in the shell; for instance:
> 
> ```
> ./solo5-hvt --net=tap100 ./app.hvt --json $(printf "%q" '{"foo": "bar"}')
> ```
> and/or we could ship a simple binary which does the right escaping; Note that I still don't have yet a global understanding of the issue here, so please bear with me if I am causing too much noise :p

I think a dedicated binary would be too much and not appropriate, because, as you are also demonstrating here, this can be "fixed" with shell-foo. (I always used 
`--json $(cat my.json | sed 's/\(["\ ]\)/\\\1/g')`
in my cases.)
So, of course we can decide that this is something the user has to deal with. Then I would at least make it very clear in the documentation that
```... ./app.hvt arg1="foo bar" arg2="foo" arg3="bar"```
is _always_ identical to
```... ./app.hvt "arg1=foo bar arg2=bar arg3=bla"```
from unikernel perspective. I would in this case go even further in managing expectations and only allow a _single command line argument_ for the unikernel (no matter if hvt or unix or whatever), because that is what several arguments are actually combined to, before they are then processed by parse-argv. And it seems we can't move away from that restriction, since some backends (xen, multiboot, ...) only allow a single string as boot time argument.

> We could also add a `file` datatype where you want to pass the contents of a file to the unikernel (e.g. `--foo=file.json`). I think this could be a nice addition (and the contents of the file could be automatically base-64 encoded by the mirage tool to avoid annoying escaping).

But that would require to even include a base64 encoder into the solo5 tender, which I can't imagine @mato would fancy. ;-) 


@hannesm:
> looking again at this issue, I don't understand the modification, as in what is supported and what not? I can see that some test cases are added, but what happens in nested quote modes (i.e. `"'foo'"` and `'"foo"'`)?

The nested quotes are handled as normal data, no escaping required.

> I'm not an expert in command-line arguments, for me that currently supported by MirageOS/parse-argv are sufficient, but I understand that escaping using backslashes is not good for all the use cases. maybe someone from the @mirage/core team interested in command-line parsing can take a deeper look? from what I understand, I now additionally need to escape my `'` in command-line arguments, but as an exchange I get `'` handling, where `"` and whitespaces may appear - is this correct?

Yes. (**Edit:** and `\` can also appear within `'`, so basically everything besides `'` itself.)

> if others are in favour of this PR, and before merging, I'd suggest:
> 
> * regarding the test cases, they are pretty hard to read due to OCaml's string escaping syntax, I'd suggest to revise them using `{|foo|}`, since there's no need to escape `"` inside of that construct.

That is a good suggestion. I wasn't aware of that possibility.

> * given that this PR changes the semantics of parse-argv drastically, I'd suggest to also document the (new) semantics in the `mli` file, and summarize it in README.md.

Agreed.


@mato:
> Yes, this is what I meant. Ignoring the JSON question, what I think what you're asking for is:
> 
> * Is there a structured way to pass arbitrary items of _data_ to a unikernel at boot time?
> 
> Where _data_ is something less restrictive than "command-line safe characters", i.e. in the general case could just be "binary blob", or at least "text/plain".

Yes, exactly.

> My argument is that the "command line" is the wrong abstraction for this. I agree that we don't have a good alternative, however, rather than trying to shoe-horn this into the "command line", perhaps we should instead consider the above as a requirement for an enhancement and design a new abstraction, taking into account the various backends and how they are booted?

That's the point. Command line is the only interface for boot time configuration across all backends. As I stated above, I would then rather prefer that - to be really consistent with the "uniformity between targets" requirement - the command line interfaces of _all_ the different backends only accept a single argument as "command line options", that would make it more transparent what is going on and might cause less confusion.

> Oh, and thinking about it, there is a simple if not super efficient solution to your use case: Just pass the data in as Base64, as defined in [RFC 4648](https://tools.ietf.org/html/rfc4648). I.e. `--json=<Base64>` and then decode it on the unikernel side. Problem solved. Note that there are some (undocumented) limits for command line length in Solo5, but I presume you're not trying to pass something larger than a couple of kB.

I am actually doing this already in some cases, like binary key data. In other cases (like the JSON example) it would be nice to just directly "cat" it from the filesystem into the command line, so that you can modify it without reencoding. Sure, like with the escaping before, a `$(cat my.json | base64)` would also work then. As so often: just because there are workarounds doesn't mean, that we couldn't improve it.

**Comment by avsm** [comment]
I'm in favour of this PR as a short-term fix.  In the long term, I agree with @ansiwen's sentiment that we should be consistent in enforcing the 'one argument allowed' across all backends.

That implies that we need to be using a format that is easier to serialise and deserialise data structures out of.  In Mirage land, that's an s-expression... so we would need to modify functoria to send a single s-expression to all boot layers, and for parse-argv to handle that appropriately (in a major version bump).

**Comment by hannesm** [comment]
6 years later, we're going to archive this repository -- development continues on https://github.com/mrage/mirage-bootvar

Since this PR conflicts with the main branch here, and is controversial, would you mind to rebase and re-develop for the mirage-bootvar repository if you are keen to push this further.

Indeed, we just introduced a hint how to deal with spaces. If there is a convincing design that works across all targets (basically, xen and virtio -- here we only get a multiboot_info block where the cmd_line is a string; also in muen this is the case as far as I know).

TL;DR: thanks for your work. This repository is soon being archived. If you like to push forward, please open at https://github.com/mirage/mirage-bootvar. Thanks a lot.
